tests/NLog.UnitTests/LogToDiagnosticsTests.cs

```diff
@@ -0,0 +1,260 @@
+        /// <param name="internalLogToDiagnostic">internalLogToDiagnostics XML attribute value. If <c>null</c> attribute is omitted.</param>
+        /// <param name="logToDiagnostic">Value of <see cref="InternalLogger.LogToDiagnostics"/> property. If <c>null</c> property is not set.</param>
+        /// <returns><see cref="TraceListener"/> instance.</returns>
+        private T SetupTestConfiguration<T>(LogLevel logLevel, Func<T> createTraceListenerFunc, bool? internalLogToDiagnostic, bool? logToDiagnostic) where T : TraceListener
+        {
+            var sb = new StringBuilder("<nlog ");
+            sb.AppendFormat("internalLogLevel='{0}'", logLevel);
+            if (internalLogToDiagnostic.HasValue)
+            {
+                sb.AppendFormat(" internalLogToDiagnostics='{0}'", internalLogToDiagnostic.Value);
+            }
+            sb.Append(">");
+            sb.AppendFormat(
+                @"<targets><target name='debug' type='Debug' layout='${{logger}} ${{level}} ${{message}}'/></targets><rules><logger name='*' level='{0}' writeTo='debug'/></rules></nlog>",
+                logLevel);
+
+            LogManager.Configuration = CreateConfigurationFromString(sb.ToString());
+
+            InternalLogger.IncludeTimestamp = false;
+
+            if (logToDiagnostic.HasValue)
+            {
+                InternalLogger.LogToDiagnostics = logToDiagnostic.Value;
+            }
+
+            var traceListener = createTraceListenerFunc();
+
+            Trace.Listeners.Clear();
+            Trace.Listeners.Add(traceListener);
+
+            return traceListener;
+        }
+
+        /// <summary>
+        /// Creates <see cref="MockTraceListener"/> instance.
+        /// </summary>
+        /// <returns><see cref="MockTraceListener"/> instance.</returns>
+        private static MockTraceListener CreateMockTraceListener()
+        {
+            return new MockTraceListener();
+        }
+
+        [Theory]
+        [InlineData(null, null)]
+        [InlineData(false, null)]
+        [InlineData(null, false)]
+        [InlineData(false, false)]
+        [InlineData(true, null)]
+        [InlineData(null, true)]
+        [InlineData(true, true)]
+        public void ShouldNotLogInternalWhenLogLevelIsOff(bool? internalLogToDiagnostics, bool? logToDiagnostics)
+        {
+            var mockTraceListener = SetupTestConfiguration(LogLevel.Off, CreateMockTraceListener, internalLogToDiagnostics, logToDiagnostics);
+
+            InternalLogger.Trace("Logger1 Hello");
+
+            Assert.Equal(0, mockTraceListener.Messages.Count);
+        }
+
+        [Theory]
+        [InlineData(true, null)]
+        [InlineData(null, true)]
+        [InlineData(true, true)]
+        public void ShouldLogToDiagnosticsWhenInternalLogToDiagnosticsIsOnAndLogLevelIsTrace(bool? internalLogToDiagnostics, bool? logToDiagnostics)
+        {
```

Comment:
I understand that we have to copy-paste in the current way the tests because we would otherwise write down all combinations. 

Still, I would like to remove this code duplication. 

I like to tryout the following lib: https://github.com/AArnott/Xunit.Combinatorial

I propose:

1. Create a new branch (because we are gonna try-out, so a new branch is a nice place for it)
2. Install https://github.com/AArnott/Xunit.Combinatorial and try it out
3. Fix that everything builds.
4 If you think it's OK, we don't need a new PR, but like to rebase the new branch into this branch. (or merge). If we got it so far, let's try the rebase.